### ISSUE TITLE
Contrib Group Application:

### ISSUE THREAD

**Issue by keiserjb** [backdrop-ops/contrib#599]
Hello and welcome to the contrib application process! We're happy to have you :)

**Please indicate how you intend to help the Backdrop community by joining this group**
* Option #1: I would like to contribute a project
* Option #2: I would like to maintain a project, but have nothing to contribute at this time
* Option #3: I would like to update documentation and/or triage issue queues
Option 1 and/or Option 3

## Based on your selection above, please provide the following information:

**(option #1) The name of your module, theme, or layout**
s3fs

## (option #1) Please note these 3 requirements for new contrib projects:

- [x] Include a README.md file containing license and maintainer information.
      You can use this example: https://raw.githubusercontent.com/backdrop-ops/contrib/master/examples/README.md
- [x] Include a LICENSE.txt file.
      You can use this example: https://raw.githubusercontent.com/backdrop-ops/contrib/master/examples/LICENSE.txt.
- [x] If porting a Drupal 7 project, Maintain the Git history from Drupal.

**(option #1 -- optional) Post a link here to an issue in the drupal.org queue notifying the Drupal 7 maintainers that you are working on a Backdrop port of their project**
[https://www.drupal.org/project/s3fs/issues/3238353](https://www.drupal.org/project/s3fs/issues/3238353)

**Post a link to your new Backdrop project under your own GitHub account (option #1)**
[https://github.com/keiserjb/s3fs](https://github.com/keiserjb/s3fs)

**If you have chosen option #2 or #1 above, do you agree to the [Backdrop Contributed Project Agreement](https://github.com/backdrop-ops/contrib#backdrop-contributed-project-agreement)**

YES

**Comment by jenlampton** [comment]
@keiserjb thank you for your application to join the contrib group! I've reviewed your module and since it meets all 3 requirements, an invite to the group is on the way! 

**(option #1 -- optional) Post a link here to an issue in the drupal.org queue notifying the Drupal 7 maintainers that you are working on a Backdrop port of their project**

[https://www.drupal.org/project/s3fs/issues/3238353](url)
**Post a link to your new Backdrop project under your own GitHub account (option #1)**

[https://github.com/keiserjb/s3fs](url)
**(option #2) If you have already contributed code to Backdrop core or contrib projects, please provide 1-3 links to pull requests or commits**

**Comment by keiserjb** [comment]
Thanks @jenlampton, I want to move forward with the Do a Barrel Roll module first because it is simpler and works all the way but there's really no harm in moving s3fs as well.  I'll look at how to do this tomorrow.

**Comment by keiserjb** [comment]
I moved Do a Barrel Roll and released it.  Hope I did it correctly.

**Comment by jenlampton** [comment]
@keiserjb looks pretty good, but the barrel roll project is missing its LICENSE.txt file ;)

**Comment by keiserjb** [comment]
dangit, well I didn't completely mess it up

**Comment by jenlampton** [comment]
nope, this is an easy fix!

**Comment by keiserjb** [comment]
I just moved s3fs.  I need to edit the About description again.  Should have done that first.

**Comment by jenlampton** [comment]
I'm also going to do a quick code review for the s3fs module, but items in this comment are only suggestions, take or leave them as you see fit. We do this review specifically looking for things that may be done differently between Backdrop and Drupal in the hopes that this information might help with the transition for those new to Backdrop :)

* Add the required sections into the README.md file. See [this example](https://github.com/backdrop-ops/contrib/blob/main/examples/README.md) for all sections that are required, but I know Maintainers and License are not optional.
    * In the License section, indicate that the bundled library uses the apache license. 
* Fix the markdown formatting in the README.md file. (This file is parsed and used to generate the project page on backdropcms.org, see the example file for formatting suggestions)
* Remove the duplicate README.txt file (only the .md version is necessary)
* in the` .info` file, the `version = 1.x` line should be removed -- this will be added automatically by the backdropcms.org packaging script when a release is created. (This makes it easier on contrib devs because you don't need to remember to update it with every release)
* in the` .info` file, the` project` line should be removed. (This line will be added by the backdrop packager once your module gets its first official release. Though it's not problematic to have it, it's not necessary since what you put on that line will be overridden by the line that is automatically added.)
* in the `s3fs.install` file, all `function s3fs_update_7*` functions can be removed. Update functions for Backdrop will start with `function s3fs_update_1`, and all drupal 7 updates should have been run in Drupal (before upgrading to Backdrop).
* in the `.module` file the `hook_help` can be removed. (That hook does not exist in Backdrop.)
* (in the `.admin.inc` file, the function `s3fs_settings` could open with ` $form = array('#config' => `s3fs.settings`);` instead of ` $form = array();`. This `#config` item allows `system_settings_form()` to save all the values from this form into the specified config file. Then, you could delete `s3fs_settings_form_submit()` function entirely.
* the `s3fs.tests.info` file should be moved inside the `tests` directory (and the line `file = tests/s3fs.test` within, should be updates to `file = s3fs.test` so that it is still relative).
* (general coding standards issue) all function definitions should end the line with the opening `{` rather than placing that character on the next line. (This will decrease line-count and keep git diffs cleaner.)
* (optional) in `s3fs.install` you can remove the blank line at the top of the file. These were necessary in Drupal for the CVS version control system, but are no longer necessary now that Drupal (&Backdrop) use Git.
* (optional) the file `S3fsStreamWrapper.inc` should be moved into an `includes` directory, since it contains a class. 
* (optional) remove the `.gitattributes` file from the repo

**Comment by keiserjb** [comment]
Thank you, @jenlampton this is all very helpful.  This one is not ready for primetime, but I know that the most basic functionality works.  I can save to an s3 bucket and manage the image files.  All of the magazine pages [https://www.library.modelaviation.com/publication/model-aviation](https://www.library.modelaviation.com/publication/model-aviation) are coming from a bucket [ https://library-modelaviation.s3.us-east-2.amazonaws.com/styles/medium/s3/ma201505_001_0.jpg](https://library-modelaviation.s3.us-east-2.amazonaws.com/styles/medium/s3/ma201505_001_0.jpg.)  I'm amazed that this works.

I have two versions of the module in the repo and I think the once that relies on the old version of aws sdk should be ditched.

**Comment by keiserjb** [comment]
A question about the barrel roll repo, I want to work on the unreleased D8 version too and make it work with 9.  I probably need to recreate a repo in my personal github and push the D9 there to keep it separate from Backdrop.

**Comment by jenlampton** [comment]
@keiserjb as long as the work is on separate branches it should be easy to push the branches to separate repos -- but there won't be any harm if the Drupal branches end up in in the backdrop repo, so I wouldn't worry about it to much. You could work off your own fork of the Backdrop repo to keep everything simple :)

**Comment by keiserjb** [comment]
I went through the list above for s3fs over the last few days.  Getting the configuration to work was seemingly more complicated than just changing the form in s3fs.admin.inc.  There was a function that turned the form_state into a configuration array and that got sent to the s3fs.module and the S3fsStreamWrapper.inc.  There was quite a lot of replacement of things like $config['use_cname'] to $config->get('s3fs_cname').

I think it is working now though.  I get it to install and configure without issue on clean Backdrop installs.  However, on sites that I've used the previous version using the old configuration style I'm getting unexpected behavior.  But I think I've figured out a work around by manually putting the json config in the active config directory.  Got to test that a few more times before I try it on a live site.
